include/knowhere/index_node.h

```diff
@@ -115,6 +124,114 @@ class IndexNode : public Object {
     Version version_;
 };
 
+struct PairComparator {
+    bool larger_is_closer;
+    PairComparator(bool larger) : larger_is_closer(larger) {
+    }
+
+    bool
+    operator()(const std::pair<float, int64_t>& a, const std::pair<float, int64_t>& b) const {
+        if (a.second == -1) {
+            return false;
+        }
+        if (b.second == -1) {
+            return true;
+        }
+        // to ensure deterministic behavior
+        if (a.first == b.first) {
+            return a.second < b.second;
+        }
+        return larger_is_closer ? a.first > b.first : a.first < b.first;
+    }
+};
+
+// Common superclass for iterators that expand search range as needed. Subclasses need
+// to override `next_batch` which will add expanded vectors to the results. For indexes
+// with quantization, override `raw_distance`.
+class IndexIterator : public IndexNode::iterator {
+ public:
+    IndexIterator(bool larger_is_closer, float refine_ratio = 0.0f)
+        : refine_ratio_(refine_ratio),
+          refine_(refine_ratio != 0.0f),
+          // std::priotity_queue is by default a max heap, so we need to reverse the comparator
+          res_(PairComparator(!larger_is_closer)),
+          refined_res_(PairComparator(!larger_is_closer)) {
+    }
+
+    std::pair<int64_t, float>
+    Next() override {
+        if (!initialized_) {
+            UpdateNext();
+            initialized_ = true;
+        }
+        auto& q = refined_res_.empty() ? res_ : refined_res_;
+        if (q.empty()) {
+            throw std::runtime_error("No more elements");
+        }
+        auto ret = q.top();
+        q.pop();
+        UpdateNext();
+        return std::make_pair(ret.second, ret.first);
+    }
+
+    [[nodiscard]] bool
+    HasNext() override {
+        if (!initialized_) {
+            UpdateNext();
+            initialized_ = true;
+        }
+        return !res_.empty() || !refined_res_.empty();
+    }
+
+ protected:
+    virtual void
+    next_batch(std::function<void(const std::vector<std::pair<float, int64_t>>&)> batch_handler) = 0;
+    // will be called only if refine_ratio_ is not 0.
+    virtual float
+    raw_distance(int64_t id) {
+        if (!refine_) {
+            throw std::runtime_error("raw_distance should not be called for indexes without quantization");
+        }
+        throw std::runtime_error("raw_distance not implemented");
+    }
+
+    const bool refine_ratio_;
+    const bool refine_;
+
+    std::priority_queue<std::pair<float, int64_t>, std::vector<std::pair<float, int64_t>>, PairComparator> res_;
```

Comment:
there are too many `std::pair<float, int64_t>` structs in the code. I'd create a simple 2-field structure with clear field names, because the current code uses `.first` or `.second` and who knows that they mean